kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/model/ingress/ClusterIPIngressDefinition.java

```diff
@@ -97,29 +107,158 @@ public Stream<ContainerPort> proxyContainerPorts() {
+    public static String bootstrapServiceName(VirtualKafkaCluster cluster, String ingressName) {
         Objects.requireNonNull(cluster);
-        Objects.requireNonNull(resource);
-        return name(cluster) + "-" + name(resource);
+        Objects.requireNonNull(ingressName);
+        return name(cluster) + "-" + ingressName;
+    }
+
+    private String qualifiedServiceHost() {
+        return qualified(cluster(), bootstrapServiceName(cluster(), name(ingress())));
     }
 
     @Override
-    public ClusterIPIngressInstance createInstance(int firstIdentifyingPort, int lastIdentifyingPort) {
-        return new ClusterIPIngressInstance(this, firstIdentifyingPort, lastIdentifyingPort);
+    public IngressModel createIngressModel(@Nullable Integer firstIdentifyingPort, @Nullable Integer lastIdentifyingPort, @Nullable Integer sharedTLSPort) {
+        ClusterIP.NodeIdentifiedBy strategy = nodeIdentifiedByStrategy();
+        switch (strategy) {
+            case PORT -> {
+                Objects.requireNonNull(firstIdentifyingPort);
+                Objects.requireNonNull(lastIdentifyingPort);
+                return new PortIdentifiesNodeClusterIPIngressModel(this, firstIdentifyingPort, lastIdentifyingPort);
+            }
+            case SNI -> {
+                Objects.requireNonNull(sharedTLSPort);
+                return new SNIIdentifiesNodeClusterIPIngressModel(this, sharedTLSPort);
+            }
+            default -> throw new IllegalStateException("Unimplemented node identification strategy: " + strategy);
+        }
     }
 
     @Override
     public int numIdentifyingPortsRequired() {
-        // one per broker plus the bootstrap
-        return nodeCount() + 1;
+        ClusterIP.NodeIdentifiedBy strategy = nodeIdentifiedByStrategy();
+        switch (strategy) {
+            case PORT -> {
+                // one per broker plus the bootstrap
+                return nodeCount() + 1;
+            }
+            case SNI -> {
+                return 0;
+            }
+            default -> throw new IllegalStateException("Unimplemented node identification strategy: " + strategy);
+        }
+    }
+
+    @Override
+    public boolean requiresSharedTLSPort() {
+        ClusterIP.NodeIdentifiedBy strategy = nodeIdentifiedByStrategy();
+        return strategy == SNI;
     }
 
     // note: we use CRD validation to enforce end >= start at the apiserver level
     private int nodeCount() {
         return nodeIdRanges.stream().mapToInt(range -> toIntExact((range.getEnd() - range.getStart()) + 1)).sum();
     }
 
+    private ClusterIP.NodeIdentifiedBy nodeIdentifiedByStrategy() {
+        return Optional.ofNullable(ingress.getSpec().getClusterIP().getNodeIdentifiedBy()).orElse(PORT);
+    }
+
+    public static String qualified(HasMetadata resource, String serviceName) {
+        return serviceName + "." + namespace(resource) + ".svc.cluster.local";
```

Comment:
Hmm `cluster.local` worming its way in again... Is the `svc` subdomain constant?